lib/src/rules/unnecessary_parenthesis.dart

```diff
@@ -67,6 +67,13 @@ class _Visitor extends SimpleAstVisitor<void> {
       return;
     }
 
+    // Constructor field initializers are rather unguarded by delimiting
+    // tokens, which can get confused with a function literal. See test cases
+    // for issues #1395 and #1473.
+    if (parent is ConstructorFieldInitializer) {
```

Comment:
Could we write something like `parent is ConstructorFieldInitializer && containsFunctionExpression(node)`? Seems like we could write a simple visitor that would visit the parenthesized expression's children looking for a closure. That should allow us to fix the bug without hiding valid issues.

Reply:
Wouldn't the visitor have to be very very complete and complex, to find specific nodes arbitrarily deep? I may be looking at it wrong... is there an example of another rule doing this?

Reply:
I'm not sure whether there are any other rules doing this, but I think the visitor should be fairly simple. I haven't tested it, but something like the following ought to work:
```
class ClosureFinder extends UnifyingAstVisitor<void> {
  bool hasClosure = false;

  @override
  void visitFunctionExpression(FunctionExpression node) {
    hasClosure = true;
  }

  @override
  void visitNode(AstNode node) {
    if (!hasClosure) {
      node.visitChildren(this);
    }
  }
}
```

Reply:
Ah, very nice! Done.